x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/stats/StatsDirectoryWrapper.java

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.index.store.stats;
+
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FilterDirectory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.LongAdder;
+import java.util.function.LongConsumer;
+import java.util.stream.Collectors;
+
+/**
+ * Directory wrapper that records various information about how {@link IndexInput} are accessed.
+ */
+public class StatsDirectoryWrapper extends FilterDirectory {
+
+    private final Map<String, LiveStats> records = ConcurrentCollections.newConcurrentMap();
```

Comment:
I should have added a precision here. The current implementation tracks "live" stats for IndexInputs, meaning that stats can be reported while they are actively updated. It allows to track progress while index inputs are used which I find interesting with IndexInput that can be slow to read. But it comes with additional burden since `LiveStats` must be used concurrently and captured on demand.

Another option would be to allow each IndexInput to update its own stats and only merge them when the IndexInput is closed.

Reply:
:+1: I think this is a good choice, because I think the updating of the stats should be a good deal cheaper than the operations for which we're collecting stats.